proposal/first-class-calc.md

```diff
@@ -0,0 +1,536 @@
+> expressions. Because interpolation could inject any part of a `calc()`
+> expression regardless of syntax, for full compatibility it's necessary to
+> parse it very expansively.
+>
+> Note that the interpolation in the definition of `CalcValue` is only reachable
+> from a `CssMinMax` production, *not* from `CalcExpression`. This is
+> intentional, for backwards-compatibility with the existing `CssMinMax` syntax.x
+
+### `CssMinMax`
+
+This proposal replaces the reference to `CalcValue` in the definition of
+[`CssMinMax`] with `CalcArgument`.
+
+[`CssMinMax`]: ../spec/syntax.md#minmaxexpression
+
+> Note that this increases the number of cases where a `MinMaxExpression` will
+> be parsed as a `CssMinMax` rather than a `FunctionExpression` (for example,
+> `min($foo, $bar)` is now a valid `CssMinMax` where it wasn't before).
+> Fortunately, this is backwards-compatible, since all such `MinMaxExpression`s
+> that were already valid will be simplified down into the same number they
+> returned before.
+
+## Types
+
+This proposal introduces a new value type known as a "calc", with the following
+structure:
+
+```ts
+interface Calc {
+  name: string;
+  arguments: CalcValue[];
+}
+
+type CalcValue =
+  | Number
+  | UnquotedString
+  | CalcOperation
+  | Calc;
+
+interface CalcOperation {
+  operator: '+' | '-' | '*' | '/';
+  left: CalcValue;
+  right: CalcValue;
+}
+```
+
+Unless otherwise specified, when this specification creates a calc, its name is
+"calc".
+
+### Operations
+
+A calc follows the default behavior of all SassScript operations, except that it
+throws an error if used as an operand of a unary or binary `+` or `-` operation.
+
+> This helps ensure that if a user expects a number and receives a calc instead,
+> it will throw an error quickly rather than propagating as an unquoted string.
+
+### Serialization
+
+#### Calc
+
+To serialize a calc, emit its name followed by "(", then each of its arguments
+separated by ",", then ")".
+
+#### `CalcOperation`
+
+To serialize a `CalcOperation`:
+
+* Let `left` and `right` be the result of serializing the left and right values,
+  respectively.
+
+* If the operator is `"*"` or `"/"` and the left value is a `CalcOperation` with
+  operator `"+"` or `"-"`, emit `"("` followed by `left` followed by `")"`.
+  Otherwise, emit `left`.
+
+* Emit `" "`, then the operator, then `" "`.
+
+* If the operator is `"*"` or `"/"` and the right value is a `CalcOperation`
+  with operator `"+"` or `"-"`, emit `"("` followed by `right` followed by
+  `")"`. Otherwise, emit `right`.
+
+## Procedures
+
+### Simplifying a Calc
+
+This algorithm takes a calc `calc` and returns a number or a calc.
+
+> This algorithm is intended to return a value that's CSS-semantically identical
+> to the input.
+
+* Let `arguments` be the result of [simplifying](#simplifying-a-calcvalue) each
+  of `calc`'s arguments.
+
+* If `calc`'s name is `"calc"`, `arguments` must contain a single argument. If
```

Comment:
This `must` here looks weird to me. It could be interpreted 2 ways:

- it triggers an error if not satisfied (the usual meaning of `must` in a spec)
- it must be valid to satisfy the `if` here.

From the explanation in the PR, I understand that you mean the second case. But that would be better written as a condition in the `if` instead.

Reply:
It actually means "the parser only allows a `calc` with name `"calc"` that has one argument". I'll make that more explicit.